scripts/nagios/check-rabbitmq-queue

```diff
@@ -31,50 +36,147 @@ states = {
+                                             universal_newlines=True)
+list_consumers_output = subprocess.check_output(['/usr/sbin/rabbitmqctl', 'list_consumers'],
+                                                universal_newlines=True)
+
+queue_counts_rabbitmqctl = dict()
+for line in list_queues_output.split("\n"):
     line = line.strip()
     m = pattern.match(line)
     if m:
         queue = m.group(1)
         count = int(m.group(2))
-        this_status = 0
-        if count > CRIT_THRESHOLD.get(queue, CRIT_THRESHOLD_DEFAULT):
-            this_status = 2
-            warn_queues.append(queue)
-        elif count > WARN_THRESHOLD.get(queue, WARN_THRESHOLD_DEFAULT):
-            this_status = max(status, 1)
-            warn_queues.append(queue)
-
-        status = max(status, this_status)
-        max_count = max(max_count, count)
-
-warn_about = ", ".join(warn_queues)
-now = int(time.time())
-now_struct = time.gmtime(now)
-
-# While we are sending digest emails, at 1800 hrs (UTC) each weekday, the mail
-# queues can get backed up; don't alert on those. Additionally, certain workers
-# (slow_queries and digest_emails) have a polling algorithm that means it's
-# normal for them to accumulate items.
-if not set(warn_queues) - set(("digest_emails", "slow_queries")) and \
-        now_struct.tm_hour == 18 and now_struct.tm_min < 25:
+        queue_counts_rabbitmqctl[queue] = count
+
+queues_with_consumers = []
+for line in list_consumers_output.split('\n'):
+    parts = line.split('\t')
+    if len(parts) >= 2:
+        queue_name = parts[0]
+        if queue_name.startswith("notify_tornado"):
+            continue
+        queues_with_consumers.append(queue_name)
+
+queue_stats_dir = subprocess.check_output([os.path.join(ZULIP_PATH, 'scripts/get-django-setting'),
+                                           'QUEUE_STATS_DIR'],
+                                          universal_newlines=True).strip()
+queue_stats = dict()  # type: Dict[str, Dict[str, Any]]
+queues_to_check = set(normal_queues).intersection(set(queues_with_consumers))
+for queue in queues_to_check:
+    fn = queue + ".stats"
+    file_path = os.path.join(queue_stats_dir, fn)
+    if not os.path.exists(file_path):
+        queue_stats[queue] = {}
+        continue
+
+    with open(file_path, "r") as f:
+        try:
+            queue_stats[queue] = json.load(f)
+        except json.decoder.JSONDecodeError:
+            queue_stats[queue] = {}
+
+results = []
+for queue_name, stats in queue_stats.items():
+    results.append(analyze_queue_stats(queue_name, stats))
+
+results.extend(check_other_queues(queue_counts_rabbitmqctl))
+
+status_list = [result['status'] for result in results]
+if 2 in status_list:
+    # CRITICAL if there's any CRITICAL in the results
+    status = 2
+elif 1 in status_list:
+    # WARN if there's any WARN in the results
+    status = 1
+elif 3 in status_list:
+    # UNKNOWN if there's any UNKNOWN in the results
+    status = 3
```

Comment:
UNKNOWN should be above `WARN` in priority, and maybe even critical; I think we could just do the maximum of the values and have that be correct.

Reply:
Fixed!